client/src/controllers/ClipboardController.ts

```diff
@@ -0,0 +1,35 @@
+import { Controller } from '@hotwired/stimulus';
+
+export class ClipboardController extends Controller<HTMLElement> {
+  static targets = ['value'];
+
+  declare valueTarget:
+    | HTMLInputElement
+    | HTMLTextAreaElement
+    | HTMLSelectElement;
+
+  copy() {
+    this.dispatch('start');
+    if (!navigator.clipboard) {
+      this.showError();
+      return;
+    }
+
+    const value = this.valueTarget.value;
+
+    if (!value) return;
+
+    navigator.clipboard.writeText(value).then(
+      () => this.showSuccess(),
+      () => this.showError(),
+    );
+  }
+
+  showError() {
+    this.dispatch('error', { detail: { type: 'error' } });
+  }
+
+  showSuccess() {
+    this.dispatch('copied', { detail: { type: 'success' } });
+  }
```

Comment:
I would probably recommend a simpler approach, where we avoid the extra methods for success/error, bringing in the suggestions from above could look something like this.

Wrapping the initial check for navigator.clipboard in a promise means we can use a Promise resolve/reject all the way through.

The dispatching events will happen in one place, then/catch which means no need to abstract these out too early.


```ts

  copy() {
    const value = this.valueTarget.value;
    if (!value) return;

    const event = this.dispatch('copy');
    if (event.defaultPrevented) return;

    new Promise((resolve, reject) => {
      if (navigator.clipboard) {
        navigator.clipboard.writeText(value).then(resolve, reject);
      } else {
        reject();
      }
    })
      .then(() => this.dispatch('copied', { detail: { type: 'success' } }))
      .catch(() => this.dispatch('error', { detail: { type: 'error' } }));
  }
```